typing/scalar.mli

```diff
@@ -0,0 +1,325 @@
+(**************************************************************************)
+(*                                                                        *)
+(*                                 OCaml                                  *)
+(*                                                                        *)
+(*                 Jacob Van Buren, Jane Street, New York                 *)
+(*                                                                        *)
+(*   Copyright 2025 Jane Street Group LLC                                 *)
+(*                                                                        *)
+(*   All rights reserved.  This file is distributed under the terms of    *)
+(*   the GNU Lesser General Public License version 2.1, with the          *)
+(*   special exception on linking described in the file LICENSE.          *)
+(*                                                                        *)
+(**************************************************************************)
+
+(** This module defines:
+    The scalar types intrinsic to the OCaml compiler, and all of the
+    primitive operations defined on them.
+
+    Overview: This module provides a comprehensive type system for scalar values in OCaml.
+
+    Type Hierarchy:
+    - Scalar.t
+      - Integral: Integer types
+        - Taggable (fits in word_size-1 bits): Int8, Int16, Int
+        - Boxable (requires boxing): Int32, Int64, Nativeint
+      - Floating: Float32, Float64
+
+    Representation Forms:
+    - Value: Tagged (small integers) or boxed representation
+    - Naked: Untagged (<=31 bits) or unboxed (>31 bits) representation
+
+    Operations:
+    - Unary: Negation, successor/predecessor, byte swap, static cast
+    - Binary: Arithmetic (add, sub, mul, div), bitwise (and, or, xor),
+             shift (lsl, asr, lsr), comparisons (integer and float)
+
+    The locality parameter tracks where boxed values are allocated.
+
+    A [Scalar.t] represents a particular OCaml type that represents a scalar value. It
+    might be tagged, boxed, or neither. There is also a type parameter for the locality of
+    the scalar value, which represents the location in which that boxed values are
+    allocated.
+
+    The important consideration is for a [Scalar.t] to represent all of the argument and
+    return types of all the primitives that we want to support. The submodules are
+    organized to make it easy for use different subsets of scalars in different places.
+    Some examples:
+
+    - Primitive arguments don't depend on the locality of their arguments, but the results
+    do.
+    - Some primitives only take integers, some take only floats, and Three_way_compare
+    takes any scalar type.
+    - The bytecode compiler wants to easily map unboxed/untagged values to their [value]
+    equivalents
+    - The middle-end wants to easily cast between any integral values using only certain
+    primitives.
+*)
```

Comment:
This comment is super helpful - thanks for writing it!  I think it can be improved even further by making the last bit more concrete: the comment says the module is organized to make it easy to represent these four things, but does't actually say how they are represented. Can you show for each bullet how the types in this module are used for these purposes?

Reply:
We think this has been done